ICE tree check: expected class ‘type’, have ‘exceptional’ (error_mark) in get_qualified_type, at tree.cc:5789

<!--
File: /tmp/im/939809D59CFD7AD72C00BF35DBE7C0FC82BBAD1B9528419BF1217BEE1E28E67A.rs
-->

snippet:
````rust
//@compile-flags: -frust-incomplete-and-experimental-compiler-do-not-use
struct A<const B: A = 1, A>;

````

Version information:

4db0cf74f64086e2bb2eff029c597239df8442e1



Command:
`/home/matthias/vcs/github/gccrs/gccrs-build/gcc/crab1 -frust-incomplete-and-experimental-compiler-do-not-use`

<details><summary><strong>Program output</strong></summary>
<p>

```
/tmp/icemaker_global_tempdir.J4KdGj8QZ6tS/rustc_testrunner_tmpdir_reporting.GipQaoH6RQc6/mvce.rs:1:19: error: failed to resolve type path segment: ‘A’
    1 | struct A<const B: A = 1, A>;
      |                   ^
crab1: internal compiler error: tree check: expected class ‘type’, have ‘exceptional’ (error_mark) in get_qualified_type, at tree.cc:5789
0x301e249 internal_error(char const*, ...)
	../../gcc/diagnostic-global-context.cc:786
0xa8cfdd tree_class_check_failed(tree_node const*, tree_code_class, char const*, int, char const*)
	../../gcc/tree.cc:9234
0xa8fde7 tree_class_check(tree_node*, tree_code_class, char const*, int, char const*)
	../../gcc/tree.h:3888
0xa8fde7 get_qualified_type(tree_node*, int)
	../../gcc/tree.cc:5789
0x1de54af build_qualified_type(tree_node*, int)
	../../gcc/tree.cc:5824
0x115cd03 Rust::Compile::HIRCompileBase::compile_constant_item(unsigned int, Rust::TyTy::BaseType*, Rust::TyTy::BaseType*, Rust::Resolver::CanonicalPath const&, Rust::HIR::Expr&, unsigned long, unsigned long)
	../../gcc/rust/backend/rust-compile-base.cc:815
0x115d784 Rust::Compile::HIRCompileBase::compile_constant_expr(Rust::Compile::Context*, unsigned int, Rust::TyTy::BaseType*, Rust::TyTy::BaseType*, Rust::Resolver::CanonicalPath const&, Rust::HIR::Expr&, unsigned long, unsigned long)
	../../gcc/rust/backend/rust-compile-base.cc:567
0x115d784 Rust::Compile::HIRCompileBase::query_compile_const_expr(Rust::Compile::Context*, Rust::TyTy::BaseType*, Rust::HIR::Expr&)
	../../gcc/rust/backend/rust-compile-base.cc:582
0x10f5371 Rust::Resolver::TypeCheckBase::resolve_generic_params(Rust::HIR::Item::ItemKind, unsigned long, std::vector<std::unique_ptr<Rust::HIR::GenericParam, std::default_delete<Rust::HIR::GenericParam> >, std::allocator<std::unique_ptr<Rust::HIR::GenericParam, std::default_delete<Rust::HIR::GenericParam> > > > const&, std::vector<Rust::TyTy::SubstitutionParamMapping, std::allocator<Rust::TyTy::SubstitutionParamMapping> >&, bool, Rust::ABI)
	../../gcc/rust/typecheck/rust-hir-type-check-base.cc:597
0x104a71c Rust::Resolver::TypeCheckItem::visit(Rust::HIR::StructStruct&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:244
0x103a985 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:57
0xfda37b Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
	../../gcc/rust/typecheck/rust-hir-type-check.cc:71
0xe8752c Rust::Session::compile_crate(char const*)
	../../gcc/rust/rust-session-manager.cc:682
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

```

</p>
</details>

--- a/gcc/testsuite/rust/compile/issue-4158.rs
+++ b/gcc/testsuite/rust/compile/issue-4158.rs
@@ -1,0 +1,3 @@
+#![feature(no_core)]
+#![no_core]
+struct A<const B: A = 1, A>; // { dg-error "invalid order for generic parameters" }

